cmd/internal/kube/injector/injector.go

```diff
@@ -0,0 +1,204 @@
+	Injector interface {
+		// Injects the given sidecar into all valid Deployment Objects and returns the result as a list of unstructured objects.
+		Inject(sidecar v1.Container) ([]*unstructured.Unstructured, error)
+		// Returns a list of namespaces that contain injectable objects.
+		// This can be used to generate other Kuberenetes objects that need to be created in the same namespace.
+		InjectableNamespaces() ([]string, error)
+	}
+	injectorImpl struct {
+		// The list of Kubernetes objects to traverse during injection. This is a list of
+		// unstructured objects because we likely won't know the type of all objects
+		// ahead of time (e.g., when reading multiple objects from a YAML file).
+		objects []*unstructured.Unstructured
+	}
+)
+
+// Constructs a new Injector with Kubernetes objects derived from the given file path.
+func FromYAML(filePath string) (Injector, error) {
+	yamlContent, err := getFile(filePath)
+	if err != nil {
+		return nil, err
+	}
+
+	// Read the YAML file into a list of unstructured objects.
+	// This is necessary because the YAML file may contain multiple Kubernetes objects.
+	// We only want to inject the sidecar into Deployment objects, but we still need to parse all resources.
+	multidocReader := kyamlutil.NewYAMLReader(bufio.NewReader(bytes.NewReader(yamlContent)))
+
+	var objList []*unstructured.Unstructured
+	for {
+		raw, err := multidocReader.Read()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			return nil, err
+		}
+
+		obj, err := fromRawObject(raw)
+		if err != nil {
+			return nil, err
+		}
+
+		result, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+		if err != nil {
+			return nil, err
+		}
+
+		objList = append(objList, &unstructured.Unstructured{Object: result})
+	}
+
+	return &injectorImpl{objects: objList}, nil
+}
+
+func (i *injectorImpl) InjectableNamespaces() ([]string, error) {
+	set := make(map[string]struct{})
+	for _, obj := range i.objects {
+		gvk := obj.GetObjectKind().GroupVersionKind()
+
+		if !isInjectable(gvk) {
+			continue
+		}
+
+		var deployment *appsv1.Deployment
+		err := runtime.DefaultUnstructuredConverter.FromUnstructured(obj.Object, &deployment)
+		if err != nil {
+			return nil, err
+		}
+
+		if deployment.Namespace == "" {
+			set["default"] = struct{}{}
+		} else {
+			set[deployment.Namespace] = struct{}{}
+		}
+	}
+
+	return sets.StringKeySet(set).List(), nil
+}
+
+func (i *injectorImpl) Inject(sidecar v1.Container) ([]*unstructured.Unstructured, error) {
+	onMap := func(obj *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+		out := obj.DeepCopy()
+		gvk := out.GetObjectKind().GroupVersionKind()
+
+		if !isInjectable(gvk) {
+			return out, nil
+		}
+
+		var deployment *appsv1.Deployment
+		err := runtime.DefaultUnstructuredConverter.FromUnstructured(out.Object, &deployment)
+		if err != nil {
+			return nil, err
+		}
+
+		injectedDeployment := deployment.DeepCopy()
+		if err != nil {
+			return nil, err
+		}
```

Comment:
I think this copy is even less apropos -- presumably the object was just allocated in FromUnstructured, so this function is the only one with access to it?